Feat: Saml Login support in Konnect (#7481)
Kong/docs.konghq.com: app/konnect/org-management/sso.md

---
title: Configure Generic SSO
---

As an alternative to {{site.konnect_saas}}’s native authentication, you can set up single sign-on (SSO) access to {{site.konnect_short_name}} using OpenID Connect or SAML. This authentication method allows your users to log in to {{site.konnect_saas}} using their IdP credentials, without needing a separate login. This topic covers configuring SSO for use with various identity providers.

> We currently support the following providers: 

Reviewer: Should this list of supported providers be placed under SAML? We really only have extended test coverage for these providers using SAML. Not that they can't work with OIDC, but we haven't tested them.

* Keycloak. 

If you want to configure Okta, please see the [Okta configuration guide](/konnect/org-management/sso/).

## Prerequisites

* {{site.konnect_short_name}} must be added to your IdP as an application
* Claims are set up in your IdP

## Set up SSO in {{site.konnect_short_name}}
{% navtabs %}
{% navtab OIDC%}
1. In [{{site.konnect_saas}}](https://cloud.konghq.com/login), click {% konnect_icon organizations %} **Organization**, and then **Auth Settings**.

1. Click **Configure provider** for **OIDC**.

1. Paste the issuer URI from your IdP in the **Issuer URI** box. 

1. Paste the client ID from your IdP in the **Client ID** box.

1. Paste the client secret from your IdP in the **Client Secret** box.

1. In the **Organization Login Path** box, enter a unique string. For example: `examplepath`.

    {{site.konnect_short_name}} uses this string to generate a custom login
    URL for your organization.

    Requirements:
    * The path must be unique *across all {{site.konnect_short_name}} organizations*.
    If your desired path is already taken, you must to choose another one.
    * The path can be any alphanumeric string.
    * The path does not require a slash (`/`).

> 1. After clicking Save, configure the SP Entity ID and Login URL on your SAML IdP.

Reviewer: Let's not forget to mention the OIDC provider needs to be enabled. Something along those lines?
```suggestion
1. After clicking Save, close the configuration dialog and click Enable on your OIDC provider.
```

    - `name`: Used as the {{site.konnect_short_name}} account's `full_name`.
    - `email`: Used as the {{site.konnect_short_name}} account's `email`.
    - `groups`: Used to map users to teams defined in the team mappings upon login.
{% endnavtab %}
{% navtab SAML %}
1. Log in to {{site.konnect_saas}}, click {% konnect_icon organizations %} **Organization**, and then select **Auth Settings**.

1. Click **Configure provider** under **SAML**. 

1. Enter the **Metadata URL** from your IdP in the **IDP Metadata URL** field.

1. In the **Login Path** field, enter the unique string that matches the one in Okta. For example: `examplepath`.

   {{site.konnect_short_name}} uses this string to generate a custom login
   URL for your organization.

   Requirements:
    * The path must be unique across all {{site.konnect_short_name}} organizations.
      If your desired path is already taken, you must choose another one.
    * The path can be any alphanumeric string.
    * The path does not require a slash (`/`).

> 1. After clicking **Save**, configure the SP Entity ID and Login URL on your SAML IdP.

Reviewer: Let's not forget to mention the SAML provider needs to be enabled. Let's add another line along those lines?
```
1. Close the configuration dialog and click **Enable** on your SAML provider.
```

## Test and apply the configuration

{:.important}
> **Important:** Keep built-in authentication enabled while you are testing IdP authentication. Only disable built-in authentication after successfully testing IdP authentication.

Test the SSO configuration by navigating to the login URI based on the organization login path you set earlier. For example: `https://cloud.konghq.com/login/examplepath`, where `examplepath` is the unique login path string set in the previous steps.

If the configuration is correct, you will see the IdP sign-in page. You can now manage your organization's user permissions entirely from the IdP application.

{% endnavtab %}
{% endnavtabs %}

> ## Troubleshooting 

Reviewer: This section also only applies to SAML. The troubleshooting guide might be placed under the SAML tab as those indications are irrelevant for OIDC.

    {{site.konnect_short_name}} uses this string to generate a custom login
    URL for your organization.

    Requirements:
    * The path must be unique *across all {{site.konnect_short_name}} organizations*.
    If your desired path is already taken, you must to choose another one.
    * The path can be any alphanumeric string.
    * The path does not require a slash (`/`).

1. After clicking Save, close the configuration dialog and click Enable on your OIDC provider.

## Test and apply the configuration

{:.important}
> **Important:** Keep built-in authentication enabled while you are testing IdP authentication. Only disable built-in authentication after successfully testing IdP authentication.

You can test the SSO configuration by navigating to the login URI based on the organization login path you set earlier. For example: `https://cloud.konghq.com/login/examplepath`, where `examplepath` is the unique login path string set in the steps above. 

If your configuration is set up correctly, you will see the IdP sign-in page.

You can now manage your organization's user permissions entirely from the IdP
application.

> ## Advanced Settings

Reviewer 2: ```suggestion
## Advanced settings
```

{% endnavtab %}
{% navtab SAML %}

The {{site.konnect_short_name}} SAML integration allows you to configure various identity providers. While technically any SAML-compliant provider can be used, the following have been verified:

* Okta 
* Azure Active Directory
* Oracle Identity Cloud Service 
* Keycloak

1. Log in to {{site.konnect_saas}}, click {% konnect_icon organizations %} **Organization**, and then select **Auth Settings**.

1. Click **Configure provider** under **SAML**. 

1. Enter the **Metadata URL** from your IdP in the **IDP Metadata URL** field.

1. In the **Login Path** field, enter the unique string that matches the one in Okta. For example: `examplepath`.

   {{site.konnect_short_name}} uses this string to generate a custom login
   URL for your organization.

   Requirements:
    * The path must be unique across all {{site.konnect_short_name}} organizations.
      If your desired path is already taken, you must choose another one.
    * The path can be any alphanumeric string.
    * The path does not require a slash (`/`).

1. After clicking **Save**, configure the SP Entity ID and Login URL on your SAML IdP.

> ## Test and apply the configuration

Reviewer 2: You can't nest headers inside tabs, they break. Notice the duplicate headings here, and the incorrect nesting: 

<img width="241" alt="Screenshot 2024-07-31 at 3 30 02 PM" src="https://github.com/user-attachments/assets/ef52e97a-ac80-432e-9fdc-833379c2da1b">

Anchors also break. For example, https://deploy-preview-7481--kongdocs.netlify.app/konnect/org-management/sso/#advanced-settings won't work, and https://deploy-preview-7481--kongdocs.netlify.app/konnect/org-management/sso/#test-and-apply-the-configuration can only take you to one of the duplicate sections.

You can take one of two approaches here:
* Don't use headings for these sections. "Test" becomes another set of steps, but without a heading.
* Each heading gets a tab group.

I'm partial to each heading having its own tab group. So it would look something like this:

> ## Set up SSO in Konnect
> tab OIDC | tab SAML
> steps
> in the OIDC tab, there's an advanced settings table that doesn't get its own heading 
> 
> ## Test and apply the configuration
> tab OIDC | tab SAML

The nice thing is, once they have selected a tab, all tabs with the same name on the page will open as well. So if they're looking at SAML, the next section will also be open to SAML.

{:.important}
> **Important:** Keep built-in authentication enabled while you are testing IdP authentication. Only disable built-in authentication after successfully testing IdP authentication.

Test the SSO configuration by navigating to the login URI based on the organization login path you set earlier. For example: `https://cloud.konghq.com/login/examplepath`, where `examplepath` is the unique login path string set in the previous steps.

If the configuration is correct, you will see the IdP sign-in page. You can now manage your organization's user permissions entirely from the IdP application.

{% endnavtab %}
{% endnavtabs %}
## Troubleshooting 

### Provider specific SAML configuration

> The following section contains provider specific information and attribute mapping tables neccessary for configuring SSO. 

Reviewer 2: ```suggestion
The following section contains provider specific information and attribute mapping tables necessary for configuring SSO. 
```

{% navtabs %}
{% navtab Azure %}

> * **OIDC uses app registration**: When adding an enterprise application, note that OIDC uses app registration.
> * **Remove namespace from claim name**: Remove the namespace from the claim name in Azure.
> * **Using groups maps to the Group ID by default**.
> * **Remove Namespace by checking **Customize** on the group claim**.

Reviewer 2: ```suggestion
* When adding an enterprise application, note that OIDC uses app registration.
* Remove the namespace from the claim name in Azure. You can do this by checking **Customize** on the group claim.
* Using groups maps to the Group ID by default.
```
This was getting repetitive. Also, points 2 and 4 were part of the same thing.

> **Attribute mapping table for Azure configuration**

Reviewer 2: ```suggestion
Attribute mapping for Azure configuration:
```
Don't really need to say "table" in a table intro.

| Azure                                       | Konnect                  |
|---------------------------------------------|--------------------------|
| Identifier (Entity ID)                      | `sp_entity_id`           |
| Reply URL (Assertion Consumer Service URL)  | `callback_url`           |
| App Federation Metadata Url                 | `idp_metadata_url`       |
| email                                       | `user.email`             |
| firstname                                   | `user.givenname`         |
| lastname                                    | `user.surname`           |
| groups                                      | `user.groups`            |
| Unique user identifier                      | `user.principalname`     |

{% endnavtab %}
{% navtab Oracle Cloud %}

> * **Set name ID format to transient**: When configuring the Name ID format in Oracle Cloud, make sure to set it to `transient`.
> * **manually upload signing certificate**: You will need to manually upload the Signing certificate from `sp_metadata_url`.
>    - `cert.pem` must use the `X509Certificate` value for signing.

Reviewer 2: ```suggestion
* When configuring the Name ID format in Oracle Cloud, make sure to set it to `transient`.
* You will need to manually upload the signing certificate from `sp_metadata_url`.
   - `cert.pem` must use the `X509Certificate` value for signing.
```

> **Attribute mapping table for Oracle Cloud configuration**

Reviewer 2: ```suggestion
Attribute mapping for Oracle Cloud configuration:
```

| Oracle Cloud                                | Konnect                  |
|---------------------------------------------|--------------------------|
| Entity ID                                   | `sp_entity_id`           |
| Assertion consumer URL                      | `callback_url`           |
| App Federation Metadata Url                 | `idp_metadata_url`       |

{% endnavtab %}
{% navtab KeyCloak %}

> * **manually upload signing certificate**: You will need to manually upload the Signing certificate from `sp_metadata_url`.
>    - `cert.pem` must use the `X509Certificate` value for signing.
> * **Find your metadata endpoint**: Go to Realm Settings in Keycloak to locate your metadata endpoint. The `sp_metadata_url` for Konnect will be:`http://<keycloak-url>/realms/konnect/protocol/saml/descriptor`

Reviewer 2: ```suggestion
* You will need to manually upload the signing certificate from `sp_metadata_url`.
   - `cert.pem` must use the `X509Certificate` value for signing.
* Go to **Realm Settings** in Keycloak to locate your metadata endpoint. The `sp_metadata_url` for {{site.konnect_short_name}} will be:`http://<keycloak-url>/realms/konnect/protocol/saml/descriptor`
```

> **Attribute mapping table for KeyCloak configuration**

Reviewer 2: ```suggestion
Attribute mapping for KeyCloak configuration:
```

| KeyCloak                                    | Konnect                  |
|---------------------------------------------|--------------------------|
| Client ID                                   | `sp_entity_id`           |
| Valid redirect URI                          | `callback_url`           |
| App Federation Metadata Url                 | `idp_metadata_url`       |

{% endnavtab %}
{% endnavtabs %}

### Authentication issues with large numbers of groups

If users are assigned a very large number of groups (over 150 in most cases), the IdP may send the groups claim in a non-standard manner, causing authentication issues. 

> To work around this limitation at the IdP, we recommend using group filtering functions provided by the IdP for this purpose. 

Reviewer 2: ```suggestion
To work around this limitation in the IdP, we recommend using group filtering functions provided by the IdP for this purpose. 
```

    * The path can be any alphanumeric string.
    * The path does not require a slash (`/`).

> 1. After clicking Save, configure the SP Entity ID and Login URL on your SAML IdP.

Reviewer: Looks like this modification should have been done on the 'Click Save' instruction for SAML, not OIDC.

    - `name`: Used as the {{site.konnect_short_name}} account's `full_name`.
    - `email`: Used as the {{site.konnect_short_name}} account's `email`.
    - `groups`: Used to map users to teams defined in the team mappings upon login.
{% endnavtab %}
{% navtab SAML %}
1. Log in to {{site.konnect_saas}}, click {% konnect_icon organizations %} **Organization**, and then select **Auth Settings**.

2. Click **Configure provider** under **SAML**. 

> 3. Enter the **Metadata URL** from your IdP in the **Identity Provider Metadata URL** field, or paste the **Metadata XML** in to the Identity **Provider Metadata XML** field.

Reviewer: Although it is supported, we won't expose the raw XML configuration in the initial release.
```suggestion
3. Enter the **Metadata URL** from your IdP in the **IDP Metadata URL** field.
```

> 4. In the **Organization Login Path** field, enter the unique string that matches the one in Okta. For example: `examplepath`.

Reviewer: ```suggestion
4. In the **Login Path** field, enter the unique string that matches the one in Okta. For example: `examplepath`.
```

   {{site.konnect_short_name}} uses this string to generate a custom login
   URL for your organization.

   Requirements:
    * The path must be unique across all {{site.konnect_short_name}} organizations.
      If your desired path is already taken, you must choose another one.
    * The path can be any alphanumeric string.
    * The path does not require a slash (`/`).

> 5. Click **Save**.

Reviewer: After clicking 'Save', we should add additional instructions to guide user to go back to their SAML IdP to configure the `SP Entity ID` and `Login URL` on the third-party system.
